### PR TITLE
Fixes #1

### LINKED ISSUE #1: Add support for dynamic arrays
`µlib` should support dynamic arrays that grow (and shrink) automatically.

### PR DESCRIPTION
**Pull request by guiomard**
WIP on dynamic arrays

### CODE REVIEW

File: src/mu-arr-int.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/* mu-str.c
```
Comment (JoshuaLeeGMIT) [inline src/mu-arr-int.c:1]: You can change this.

File: src/mu-arr-int.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <stdlib.h>
+
+#include "mu-arr-int.h"
+
+static MuArrInt * 
+mu_arrint_init ()
```
Comment (JoshuaLeeGMIT) [inline src/mu-arr-int.c:24]: It shouldn't be a static function or it won't be visible outside the file. We should probably follow the same naming style, so something like:

`mu_arr_int_new ()`

**Comment by JoshuaLeeGMIT** [comment]
Nice stuff, thanks Gavin!

The first thing is it'd probably be easier for us to make it more generic. When we create a new array, we should be able to specify the size of what we're going to store. Then we can store the element size in the struct and use that when expanding.

```
MuArr *
mu_arr_new (size_t size)
{

}
```

Then if I want an `int` array I can just do `arr = mu_arr_new (sizeof (int));` and I can do my own objects like `arr = mu_arr_new (sizeof (MyStruct));`.

File: src/mu-arr-int.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/* mu-str.h
```
Comment (JoshuaLeeGMIT) [inline src/mu-arr-int.h:1]: Same here.

File: src/mu-arr-int.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#pragma once
+
+typedef struct _MuArrInt MuArrInt;
+
+struct _MuArrInt
```
Comment (JoshuaLeeGMIT) [inline src/mu-arr-int.h:23]: We probably need to store the length of the array as well as the element size. If we go more generic, we should try having the array be `void *`.

**Review by JoshuaLeeGMIT — COMMENTED** [review]
You should add the new files to the `meson.build` file in `src` so it builds too.

File: src/mu-arr-int.c (hunk 2 above)
Comment (JoshuaLeeGMIT) [reply src/mu-arr-int.c:24]: Ah, sorry, I see what you were going for now. Should we keep the `_new/_add/_append` split?

**Comment by guiomard** [comment]
Don't worry this was just a first run through, thanks a lot for the feedback. I was planning to make it a generic array later, int was really just a placeholder. I'm not sure what you mean by "Should we keep the _new/_add/_append split?" though. I kept the init function static to make it as streamlined as possible but maybe it should just be accessible anyways. I'll rename it to new regardless.
What do you think about storing the data type as a pointer rather than an explicit array?

**Comment by JoshuaLeeGMIT** [comment]
>  I'm not sure what you mean by "Should we keep the _new/_add/_append split?" though

I mean, I'd prefer to keep creating our objects the same way. To get a string we use `mu_str_new()` so I think we should get an array my using `mu_arr_new (sizeof (Whaterver))` so that initilialisation is handled in that function when the caller requests a new array and we don't need a separate init function.

Something like:

```C
MuArr *
mu_arr_new (size_t size)
{
  MuArr *arr;

  arr = malloc (sizeof (MuArr));

  arr->array = malloc (size * SOME_DEFAULT);
  arr->len = 0;
  arr->el_size = size;

  return arr;
}
```

> What do you think about storing the data type as a pointer rather than an explicit array?

I think we have to. It'll have to be `malloc`'d and then `realloc`'d as it grows anyway. I was wrong about using `void *` too, we can't do pointer arithmetic on void pointers so it'll have to be `char *`.

https://stackoverflow.com/questions/3922958/void-arithmetic

**Comment by JoshuaLeeGMIT** [comment]
@G00358077 I'd say we should merge this tonight for the review tomorrow as well as my linked lists stuff. We can work on it again.

**Pull request merged**